app/src/lib/stores/app-store.ts

```diff
@@ -3847,18 +3862,66 @@ export class AppStore extends TypedBaseStore<IAppState> {
     this.hasUserViewedStash = false
   }
 
-  private async refreshAfterCheckout(repository: Repository, branch: Branch) {
+  /**
+   * @param commitish A branch name or a commit hash
+   */
+  private async refreshAfterCheckout(
+    repository: Repository,
+    commitish: string
+  ) {
     this.updateCheckoutProgress(repository, {
       kind: 'checkout',
       title: `Refreshing ${__DARWIN__ ? 'Repository' : 'repository'}`,
+      description: 'Checking out',
       value: 1,
-      targetBranch: branch.name,
+      target: commitish,
     })
 
     await this._refreshRepository(repository)
     return repository
   }
 
+  /**
+   * Checkout the given commit, ignoring any local changes.
+   *
+   * Note: This shouldn't be called directly. See `Dispatcher`.
+   */
+  public async _checkoutCommit(
+    repository: Repository,
+    commit: CommitOneLine
+  ): Promise<Repository> {
+    const repositoryState = this.repositoryStateCache.get(repository)
+    const { branchesState } = repositoryState
+    const { tip } = branchesState
+
+    // No point in checking out the currently checked out commit.
+    if (tip.kind === TipState.Valid && tip.branch.tip.sha === commit.sha) {
+      return repository
+    }
```

Comment:
I guess we could also check for `tip.kind === TipState.Detached` and see if `tip.currentSha === commit.sha`, right? 🤔 

```suggestion
    if (
      (tip.kind === TipState.Valid && tip.branch.tip.sha === commit.sha) ||
      (tip.kind === TipState.Detached && tip.currentSha === commit.sha)
    ) {
      return repository
    }
```

Reply:
@kitswas This lead me to check what happens when you try to checkout the already checked out commit. Definitely, we need to add this check. But, also, we need add this check in the history log, because right now you can invoke the context menu on the first commit and the check out commit is enabled. I click it and get the confirm, then nothing happens. It should be disabled for the first commit since you cannot check that one out/it is already checked out.